modules, backend: init

tupakkatapa opened the pull request:
Creating a module for hosting the backend with an SSL endpoint. Probably initializing the `backend-demo` host in or during this PR as well, since this needs to be tested and likely iterated on. 

This should be rebased from `main` after #64 is merged.

jhvst commented on packages/backend/src/main.rs:
```diff
@@ -224,11 +224,24 @@ async fn main() -> std::io::Result<()> {
     let addr = matches.get_one::<String>("addr").unwrap();
     let port = matches.get_one::<String>("port").unwrap();
-    let base_url = format!("http://{}:{}", addr, port);
+    let listen = format!("{}:{}", addr, port);
+    let domain = matches.get_one::<String>("domain");
+    let base_url = if let Some(domain) = domain {
+        format!("https://{}", domain)
+    } else {
+        format!("http://{}", listen)
+    };
```
A note: according to the documentation, `format!` macro panics if it fails. It's probably fine given its done before the main loop of the program, but you should prefer actual String methods than over-utlizing macros. Especially the code in 241 and 243 are just string concatenations. There is thus no reason to leverage type-polymorphic macros especially when those macros might crash the program.
That is, Rust gives you very ergonomic and strong type system, you should use it!

jhvst replied:
Furthermore, which is a real issue, is that the domain is unused. As you are already using Caddy as a proxy, there should be no need to run the actual Rust server through HTTPs -- either Caddy provides it, or you configure the web framework to use the certificates and remove Caddy.

jhvst replied:
I also have a feeling that the whole HTTPs certification in this particular case is moot, because the certificate for the publicly resolvable build server will receive a certificate from CloudFlare. And if someone is developing the thing locally, I think they will just run the build server without a certificate.
The ACME certificates will only be relevant on the nodes that run CL + EL, because those will need it to comply with security rules of browsers when being accessed from the hosted frontend UI.

tupakkatapa replied:
Domain is for the non-relative download links, any suggestions how should we provide it in the HTTP response otherwise?

jhvst replied:
I see, one option is to handle it on the frontend: the build server returns a relative path but the frontend appends it to the URI given on the frontend. I think this is an assumption we can make. Someone could always provide a domain (private or local) on the frontend, or then an IP address.

tupakkatapa replied:
That sounds like a smart way to handle this. To be further discussed with @EeroKokkonen and @Lagical. 

I'll revert the changes regarding this.

tupakkatapa commented:
Couldn't get this to work without server-side self-signed certificates, but the issue was actually with Caddy - it likely had some redirect rule by default, causing `ERR_TOO_MANY_REDIRECTS`. I won’t bother investigating it further right now. Switched to Nginx and removed ACME entirely from the module.

TL;DR: This now works with Cloudflare certs only.

tupakkatapa commented:
After deploying the `homestakeros-backend` host, the first build appears to return a 504 error. I will investigate and fix this tomorrow. After that, it should be ready for final review before being merged.

jhvst commented on packages/backend/src/main.rs:
```diff
@@ -39,8 +40,22 @@ async fn nixos_config(config: web::Json<Config>, data: web::Data<AppState>) -> i
     // Use the provided output directory.
     let output_dir = data.temp_dir.path();
 
+    // Create a unique working directory inside output_dir for this build.
+    let unique_id = SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .map(|d| d.as_nanos().to_string())
+        .unwrap_or_else(|_| "default".to_string());
+    let working_dir = output_dir.join(format!("build_work_{}", unique_id));
```
Like I said before, unwraps are not good. Similarly, there is again a format! used for string concatenation. The comment above the lines makes me think this is coming from an LLM, but you should check the code for unwraps and unsafe formats even if it passes compilation.

I think a much better approach would be to create a struct for the folders you want, and make sure that this struct can be created. Currently, there is a lot of code just creating directories in an ad-hoc way, which makes it very hard to review what is going on, because files are created "on-the-fly" as needed. There is no code to test these functions, so I do not know if these work. 

The correct workflow would be to separate these into functions, create a tests for them, and only if they pass your tests then call it from these functions. The current approach decreases coverage in each commit, which is the opposite what we want.

Most of the time should be spent on writing tests, because otherwise you will have to do all these interactively and spend more time on making everything works. It's very hard to maintain or rewrite parts of the code if there are no existing tests to make sure adding or removing code still retains the properties we want.

tupakkatapa replied:
Done. I'll do a quick refactoring PR next to remove some poor error handling and other unsafe practices.

tupakkatapa commented:
Tests:
```
    Running tests/functions.rs (target/debug/deps/functions-feb349e9b1fa569f)

running 3 tests
test test_compute_sha256 ... ok
test test_write_default_nix ... ok
test test_create_tarball ... ok

test result: ok. 3 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running tests/json_parsing.rs (target/debug/deps/json_parsing-75e8168680a68382)

running 7 tests
test test_invalid_types ... ok
test test_unknown_nested_fields ... ok
test test_missing_required_fields ... ok
test test_missing_nested_required_fields ... ok
test test_invalid_json ... ok
test test_valid_config ... ok
test test_unknown_fields ... ok

test result: ok. 7 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running tests/workspace.rs (target/debug/deps/workspace-f1d7e30fb2c7eb9f)

running 2 tests
test test_build_workspace_creation ... ok
test test_cleanup ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
```

jhvst commented on packages/backend/src/main.rs:
```diff
@@ -233,6 +251,8 @@ async fn main() -> std::io::Result<()> {
     println!("Running on: {}", base_url);
     let temp_dir = TempDir::new().expect("Failed to create temporary directory");
     println!("Using temporary directory: {}", temp_dir.path().display());
+    // Create a dedicated folder for final build artifacts.
+    fs::create_dir_all(temp_dir.path().join("builds"))?;
```
Can this also be part of the BuildWorkspace? It seems that the two are related.

tupakkatapa replied:
The workspace creates directories and an output link path for each individual build. The main working directory, which is the temporary directory itself, along with the builds subdirectory, only needs to be created once. Therefore, it is not directly related to this process per se. However, I can create a separate "init" function that will initialize the main directories and but should only be called once when the program starts.

jhvst replied:
Alternative option is to create a `::new()` which initializes the struct. Suppose it is bound to `ws`. This singleton is passed/accessed from the web app such that it `ws.build` or something like that then creates a new structure using struct internal fields to populate the base directory. See, e.g.,: https://refactoring.guru/design-patterns/singleton/rust/example

This might sound like overengineering, but I think it would be important to understand the relation between some of this imperative code. They are directly related in the sense that the workspace folder is a direct outcome of functions of form `f (g x)` where `f` and `g` are functions and `x` is a variable. Currently, the `g x` (tempdir creation) is ran "somewhere" thus has a disjoint relation to `f` in the sense that you must _eventually_ pass some variable to `f`, and currently, you must know its the temporary directory created before. A more logical way would be to bind these two function calls together through a struct which preserves the relation between the two. All this code will then make testing more simple as well.

jhvst commented on packages/backend/tests/workspace.rs:
```diff
@@ -0,0 +1,40 @@
+
+#[test]
+fn test_cleanup() -> Result<(), Box<dyn std::error::Error>> {
+    let temp_dir = TempDir::new()?;
+    let output_dir = temp_dir.path();
+    let hostname = "testhost";
```
I feel like these should be inside of `BuildWorkspace::new`, because it always uses TempDir.

tupakkatapa commented on packages/backend/src/workspace.rs:
```diff
@@ -0,0 +1,50 @@
+
+impl BuildWorkspace {
+    /// Creates a new workspace with a unique working directory.
+    pub fn new(output_dir: &Path, hostname: &str) -> io::Result<Self> {
+        let unique_id = match SystemTime::now().duration_since(UNIX_EPOCH) {
+            Ok(duration) => duration.as_nanos().to_string(),
+            Err(_) => "default".to_string(),
+        };
+        let mut dir_name = String::with_capacity("build_work_".len() + unique_id.len());
+        dir_name.push_str("build_work_");
+        dir_name.push_str(&unique_id);
+        let working_dir = output_dir.join(dir_name);
+
+        // Create the directories.
+        fs::create_dir_all(&working_dir)?;
+        let nix_config_dir = working_dir.join("nixConfig");
+        fs::create_dir_all(&nix_config_dir)?;
+        let hostname_dir = nix_config_dir.join("nixosConfigurations").join(hostname);
+        fs::create_dir_all(&hostname_dir)?;
+
+        // Construct a path for the nix build result.
+        let out_link = working_dir.join("kexecTree");
+
+        Ok(BuildWorkspace {
+            working_dir,
+            nix_config_dir,
+            hostname_dir,
+            out_link,
+        })
+    }
+
+    /// Cleans up the original directories.
+    pub fn cleanup(&self) -> io::Result<()> {
+        fs::remove_dir_all(&self.nix_config_dir)?;
+        fs::remove_file(&self.out_link)?;
+        Ok(())
+    }
+}
```
I will look into these, appreciating the informative reviews, since learning on the go.

jhvst reviewed (approved):
The comments I have are more about refactoring now, so feel free to alternatively squash and merge and handle the comments in a separate refactoring PR.
